app/src/main/java/org/sopt/at/presentation/auth/signup/SignUpViewModel.kt

```diff
@@ -0,0 +1,70 @@
+package org.sopt.at.presentation.auth.signup
+
+import androidx.lifecycle.ViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import org.sopt.at.model.LoginUser
+
+class SignUpViewModel: ViewModel() {
+
+    private val _currentStep = MutableStateFlow(SignUpStep.ID)
+    val currentStep: StateFlow<SignUpStep>
+        get() = _currentStep.asStateFlow()
+
+    private val _userInfo = MutableStateFlow<LoginUser>(LoginUser(id = "", password = ""))
+    val userInfo: StateFlow<LoginUser>
+        get() = _userInfo.asStateFlow()
+
+    fun updateStep(step: SignUpStep) {
+        _currentStep.value = step
+    }
+
+    fun updateId(id: String) {
+        _userInfo.update {
+            it.copy(id = id)
+        }
+    }
+
+    fun updatePassword(password: String) {
+        _userInfo.update {
+            it.copy(password = password)
+        }
+    }
+
+    fun isValidId(): Boolean {
+        return checkIdValidation(_userInfo.value.id)
+    }
+
+    fun isValidPassword(): Boolean {
+        return checkPasswordValidation(_userInfo.value.password)
+    }
+
+    companion object {
+        // 정규식 패턴
+        const val ID_PATTERN = "^(?=.*[a-z]+)(?=.*\\d*)[a-z\\d]{6,12}$"
+        const val PWD_PATTERN = "^.*(?=^.{8,15}$)(?=.*\\d)(?=.*[a-zA-Z])(?=.*[~!@#$%^&*]).*$"
+
+        fun checkIdValidation(id: String): Boolean {
+            return id.matches(ID_PATTERN.toRegex())
+        }
```

Comment:
작은 차이긴 한데
id.matches(ID_PATTERN.toRegex()) 할 때마다 regex객체를 새로 만들잖아요.
```suggestion
        val ID_PATTERN = Regex("^(?=.*[a-z]+)(?=.*\\d*)[a-z\\d]{6,12}$")
        
        fun checkIdValidation(id: String): Boolean {
            return id.matches(ID_PATTERN)
        }
```
이렇게 하면 매번 객체를 만들지 않아도 될 것 같습니다.

Reply:
그리고 Companion Object로 만드시면 저번에 말씀드린 싱글톤 패턴이 아니게 됩니다. 정규식은 CPU연산이 큰 작업이라서ㅎㅎ
디자인 패턴중 싱글톤, 빌더 패턴을 알아보시면 좋아요~

Reply:
오 생각하지 못했던 부분인데 regex 객체를 매번 만드는 것보다 훨씬 좋은 방법인 거 같네요. 짚어주셔서 감사합니다!!